fix: add usage tracking for /v1/messages endpoint

colesmcintosh opened the pull request:
## Title

fix: add usage tracking for /v1/messages endpoint

## Relevant issues

Fixes #12132

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

🐛 Bug Fix

## Changes

This PR adds usage tracking for non-Anthropic models on the `/v1/messages` endpoint by adding the missing `post_call_success_hook` call.

### Problem
The `/v1/messages` endpoint was not tracking usage for non-Anthropic models, which meant that costs and usage statistics were not being recorded when using this endpoint with models like OpenAI, Gemini, etc.

### Solution
- Added `post_call_success_hook` call for non-streaming responses to enable usage tracking
- Ensured headers are properly updated after the hook call (since the hook may modify hidden_params)
- Maintained the existing flow for streaming responses

### Testing
The existing test suite for the endpoint continues to pass:
```
tests/test_litellm/proxy/anthropic_endpoints/test_endpoints.py::TestAnthropicEndpoints::test_async_data_generator_anthropic_dict_handling PASSED
```

This change follows the same pattern used in other endpoints like `/v1/chat/completions`, ensuring consistency across the codebase.

colesmcintosh committed:
fix: merge team callbacks with config callbacks during request processing

- Modified add_litellm_data_to_request to merge callbacks instead of overwriting
- Team callbacks (from UI) now combine with existing config callbacks
- Added deduplication to prevent duplicate callbacks
- Handles both list and non-list callback values gracefully

Fixes #12118

colesmcintosh committed:
fix: add usage tracking for /v1/messages endpoint

Enable usage tracking for non-Anthropic models on the /v1/messages endpoint by adding the missing post_call_success_hook call. This ensures that usage and costs are properly tracked for all models, not just Anthropic models.

Fixes #12132

ishaan-jaff commented on litellm/proxy/litellm_pre_call_utils.py:
```diff
@@ -725,8 +725,22 @@ async def add_litellm_data_to_request(  # noqa: PLR0915
     )
     if callback_settings_obj is not None:
-        data["success_callback"] = callback_settings_obj.success_callback
-        data["failure_callback"] = callback_settings_obj.failure_callback
```
please can this pr just be limited to changes for anthropic messages

colesmcintosh replied:
Apologies - was from a separate branch

ishaan-jaff reviewed (changes requested):
-

colesmcintosh committed:
fix: limit PR scope to Anthropic messages endpoint only

Per PR feedback from @ishaan-jaff, removing unrelated team callback merging changes:
- Reverted litellm_pre_call_utils.py team callback merging logic
- Removed test_team_callback_merging.py test file
- Kept only the core fix: adding usage tracking for /v1/messages endpoint

This keeps the PR focused solely on fixing issue #12132 (missing usage tracking)

krrishdholakia commented on litellm/proxy/anthropic_endpoints/endpoints.py:
how does this solve the problem?

krrishdholakia replied:
no the logging happens in wrapper_async - which is triggered when calling the completions bridge. 

The issue was that the usage chunk was not being returned to the user

krrishdholakia commented:
Closing in favor of https://github.com/BerriAI/litellm/pull/12184
